core/cat/mad_hatter/decorators/endpoint.py

```diff
@@ -0,0 +1,102 @@
+from typing import Callable
+from fastapi import APIRouter
+
+cheshire_cat_api = None
+
+# class to represent a @endpoint
+class CustomEndpoint:
+    def __init__(self, prefix: str, path: str, function: Callable, **kwargs):
+        self.prefix = prefix
+        self.path = path
+        self.function = function
+        self.name = self.prefix + self.path
+
+        for k in kwargs:
+            setattr(self, k, kwargs[k])
+
+    def __repr__(self) -> str:
+        return f"CustomEndpoint(path={self.name})"
+
+# Called from madhatter to inject the fastapi app instance
+def _init_endpoint_decorator(new_cheshire_cat_api):
+    global cheshire_cat_api
+
+    cheshire_cat_api = new_cheshire_cat_api
+
+# @endpoint decorator. Any function in a plugin decorated by @endpoint will be exposed as FastAPI operation
+def endpoint(path, methods, prefix="/custom_endpoints", tags=["custom_endpoints"], **kwargs) -> Callable:
+    """
+    Define a custom API endpoint, parameters are the same as FastAPI path operation.
+    Examples:
+        .. code-block:: python
+            @endpoint(path="/hello", methods=["GET"])
+            def my_endpoint() -> str:
+                return {"Hello":"Alice"}
+    """
+
+    global cheshire_cat_api
+
+    def _make_endpoint(endpoint):
+        custom_endpoint = CustomEndpoint(prefix=prefix, path=path, function=endpoint, **kwargs)
+
+        plugins_router = APIRouter()
```

Comment:
This creation for each decorated function, can create a problem for fast-api in terms of management?
Is possible to create unique router that is shared for the specific plugin?

Reply:
Yep maybe one router for all of them, but I need to play a little with the PR

Also I would go for one decorator but with dot syntax 
`@endpoint.post`
`@endpoint.get` 
and so on, similar as fastapi routing (all our endpoints are written like this)

Reply:
> can create a problem for fast-api in terms of management?

There shouldn't be any problems, the automatic tests creates four `APIRouter`, we can add additional tests on this

> Is possible to create unique router that is shared for the specific plugin?

Possible but more complicated to managed and IMHO less flexible.

Reason: all the routes must be add to the router before it get included in the FastAPI app.
The method `include_router` clones all the routes inside the parent route ([see here](https://github.com/fastapi/fastapi/blob/d4ab06a2b6b59bd264c0cc43367c5a68174e8f88/fastapi/routing.py#L1302)).

This code doesn't work:
``` python
plugins_router = APIRouter()
cheshire_cat_api.include_router(plugins_router, prefix=prefix)

plugins_router.add_api_route(
    path=path, endpoint=endpoint, methods=methods, tags=tags, **kwargs
)

```